pkg/endpoint/policy.go

```diff
@@ -334,6 +334,18 @@ func (e *Endpoint) setDesiredPolicy(datapathRegenCtxt *datapathRegenerationConte
 			// Do nothing if e.policyMap was not initialized already
 			if e.policyMap != nil && e.desiredPolicy != e.realizedPolicy {
 				e.desiredPolicy.Detach(e.getLogger())
+				// Make sure e.desiredPolicy still reflects the
+				// desired state of the world after reverting
+				// back to the last known good state. On the
+				// next regeneration attempt, res.endpointPolicy
+				// will not be recalculated, since we already
+				// updated e.nextPolicyRevision; we want to
+				// preserve the desired policy calculated in
+				// this round so that we don't just end up with
+				// an empty map. See GH-38998.
+				defer func(p *policy.EndpointPolicy) {
```

Comment:
I understand why you added this, but this defer (in a revert function!) makes it much, much harder to reason about. Can we fix this by changing the policy calculation logic instead? Something to prevent short-circuiting policy calculation if `e.desiredPolicy` has the wrong revision?

Reply:
Note that `e.desiredPolicy.Detach()` has been called, so it is no longer getting incremental policy updates. I'm pretty sure we can't re-use it at this point.

Reply:
In the latest push, I swapped this out for a call to `e.forcePolicyComputation()`.